pkg/counter/cache.go

```diff
@@ -0,0 +1,166 @@
+package counter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var cachedCounts = cache{
+	Counts: Counts{
+		UserAccountsPerClusterCounts: map[string]int{},
+	},
+}
+
+// Counts is type that contains number of MURs and number of UserAccounts per member cluster
+type Counts struct {
+	MasterUserRecordCount        int
+	UserAccountsPerClusterCounts map[string]int
+}
+
+type cache struct {
+	Counts
+	sync.RWMutex
+	initialized bool
+}
+
+func write(operation func()) {
+	cachedCounts.Lock()
+	defer cachedCounts.Unlock()
+	operation()
+}
+
+// Reset resets the cached counter - is supposed to be used only in tests
+func Reset() {
+	write(func() {
+		reset()
+	})
+}
+
+func reset() {
+	cachedCounts.Counts = Counts{
+		UserAccountsPerClusterCounts: map[string]int{},
+	}
+	cachedCounts.initialized = false
+}
+
+// IncrementMasterUserRecordCount increments the number of MasterUserRecord in the cached counter
+func IncrementMasterUserRecordCount() {
+	write(func() {
+		cachedCounts.MasterUserRecordCount++
+	})
+}
+
+// DecrementMasterUserRecordCount decreases the number of MasterUserRecord in the cached counter
+func DecrementMasterUserRecordCount(log logr.Logger) {
+	write(func() {
+		if cachedCounts.MasterUserRecordCount != 0 || !cachedCounts.initialized {
+			cachedCounts.MasterUserRecordCount--
+		} else {
+			log.Error(fmt.Errorf("the count of MasterUserRecords is zero"),
+				"unable to decrement the number of MasterUserRecords")
+		}
+	})
+}
+
+// IncrementUserAccountCount increments the number of UserAccount fo the given member cluster in the cached counter
+func IncrementUserAccountCount(clusterName string) {
+	write(func() {
+		cachedCounts.UserAccountsPerClusterCounts[clusterName]++
+	})
+}
+
+// DecrementUserAccountCount decreases the number of UserAccount fo the given member cluster in the cached counter
+func DecrementUserAccountCount(log logr.Logger, clusterName string) {
+	write(func() {
+		if cachedCounts.UserAccountsPerClusterCounts[clusterName] != 0 || !cachedCounts.initialized {
```

Comment:
same question here
```suggestion
		if cachedCounts.UserAccountsPerClusterCounts[clusterName] != 0 && cachedCounts.initialized {
```

Reply:
and the same answer, but this time it applies on UserAccounts :-)

Reply:
sure, and same concern ;)